### DISCUSSION TITLE
建议调整一下参数类型

### DISCUSSION THREAD

**Discussion by xiaxiangjun** [Tencent/tquic#67] [Ideas（新功能的建议和讨论）]
```rust
fn on_packets_send(&self, pkts: &[(Vec<u8>, PacketInfo)]) -> tquic::Result<usize> {}
```
建议将 `Vec<u8>`类型改为 `bytes::Bytes`类型；这样应用层在处理时可以减少内存的copy

**Comment by iyangsj** [comment]
@xiaxiangjun `pkts`代表了加密的出向报文，应用程序发送时在用户态并不需要再进行拷贝

https://doc.rust-lang.org/std/net/struct.UdpSocket.html#method.send_to
https://docs.rs/socket2/latest/socket2/struct.Socket.html#method.send_to
https://docs.rs/libc/latest/libc/fn.sendmmsg.html

**Reply by xiaxiangjun** [reply]
用户态如果有多重架构, 例如在非标准的socket传输时，是存在copy的可能性。 用 `Vec<u8>`与用 `bytes::Bytes`对于库来说都一样方便，但是对于应用层来说就不一样了

**Reply by iyangsj** [reply]
@xiaxiangjun 谢谢你的关注和进一步的解释。明确一下， `on_packets_send` 代表应用程序发送最终加密的QUIC报文（而不是应用程序向QUIC层写入应用数据）

在应用程序使用非标准socket封装库来发送报文的场景，有可能A用户使用的是 Bytes类型，而B用户使用的是 Vec<u8>类型。无论如何处理，都会有一方发现不适配。关键在于，如果应用程序真正在意减少拷贝次数，那么其封装的socket库，应参考官方库或广泛采用的第三方库，不对类型进行不必要的限制，导致额外的拷贝。

**Comment by xiaxiangjun** [comment]
https://docs.rs/s2n-quic/latest/s2n_quic/stream/struct.BidirectionalStream.html#method.send

 s2n这个库发送的时候也是用的`bytes::Bytes`，用起来就比较方便

**Reply by iyangsj** [reply]
这个链接中 `send()`代表的是 应用程序向 QUIC 双向流写入数据 (与上面讨论的场景并不相同）
TQUIC对应接口参见：https://docs.rs/tquic/latest/tquic/connection/struct.Connection.html#method.stream_write

<img width="315" alt="image" src="https://github.com/Tencent/tquic/assets/780807/1052dc84-4e1d-4ec4-a21d-5b01c77cea92">

**Comment by xiaxiangjun** [comment]
我在集成的时候，遇到`async`问题，需要传递到其它支持`async`的地方才能正常的发送出去；用的是非标准的底层`UdpSocket`, 用的Turn的传输库。所以建议`Vec<u8>`改为`bytes::Bytes`对于底层库来说性能没有差别。
图中`buf.clone()`就会产生一次内存Copy, 如果是`bytes::Bytes`类型的话，就不会产生一次内存Copy

`fn on_packets_send(&self, pkts: &[(Vec<u8>, PacketInfo)]) -> tquic::Result<usize> {}`既然提供了这个方法，代表着发送数据的权力交给了用户处理，那么，对于使用的用户来说传输的可选择的方式就非常的多，所以如果将这里的`Vec<u8>`改为`bytes::Bytes`的话，可以适配更多的情况
